exporters/prometheus/exporter.go

```diff
@@ -0,0 +1,186 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.18
+// +build go1.18
+
+package prometheus // import "go.opentelemetry.io/otel/exporters/prometheus"
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+)
+
+// exporter is a Prometheus exporter.
+type exporter struct {
+	reader metric.Reader
+}
+
+// config is added here to allow for options expansion in the future.
+type config struct{}
+
+// Option may be used in the future to apply options to a Prometheus Exporter config.
+type Option interface {
+	apply(config) config
+}
+
+// NewPrometheusExporter returns a Prometheus exporter.
+func NewPrometheusExporter(reader metric.Reader, options ...Option) (prometheus.Collector, error) {
```

Comment:
I don't think we should be accepting a Reader here, but instead have the reader embed its own created `ManualReader`. The important part is because the Reader needs to only accept cumulative temporality to be compatible with the Prometheus exposition format. Also, the intended API was going to be something like this:

```go
tp := NewMeterProvider(WithReader(prometheus.New()))
// ...
```

Reply:
Good point.  +1

Reply:
I tried adding something like this in https://github.com/open-telemetry/opentelemetry-go/pull/3135/commits/0a782e0c92491edb29a5757f12ae9ab86a3f7dc2, ptal. The 2 `Collect()` methods meant I had to add a new struct for the prometheus Collector interface in order to have the exporter just work with a MeterProvider.

Also, will users be able to initialize it like you had? They'll still need to keep the exporter in a variable to pass it to `prometheus.NewRegistry` in practice I think

Reply:
Should the `New` function take a `WithRegisterer` option and register with that at startup. And if none is passed use the `DefaultRegisterer`? It seems like we did something similar before.

https://github.com/open-telemetry/opentelemetry-go/blob/0e6f9c29c10d6078e8131418e1d1d166c7195d61/exporters/prometheus/prometheus.go#L64-L81

Thoughts?

Reply:
That would force us to return an error from `New`, which it is nice we don't have to do right now.  It would also mean we have to support promhttp.HandlerOpts, which I don't love either.  Overall, I don't think it is that much more boilerplate to just return the prometheus.Collector interface.  Its unfortunate that we have two different Collect functions :).

Reply:
Yeah, sounds good. We could always add a `MustRegister` function in a future release that would inline the creation and registration. This wouldn't close the door.

Reply:
Yeah the old exporter had that Registerer pattern, so I originally did something like that but @dashpole suggested we could cut it out in https://github.com/open-telemetry/opentelemetry-go/pull/3135#discussion_r966237373

If this looks good as it is then I'll squash down and get it ready to merge